src/hnsw/extra_dirtied.c

```diff
@@ -0,0 +1,78 @@
+#include <postgres.h>
+
+#include "extra_dirtied.h"
+
+#include <assert.h>
+
+ExtraDirtiedBufs *extra_dirtied_new(Relation index_rel)
+{
+    ExtraDirtiedBufs *ed = palloc0(sizeof(ExtraDirtiedBufs));
+    assert(ed != NULL);
+    ed->index_rel = index_rel;
+    ed->EXTRA_DIRTIED_BLOCKNO = palloc0(sizeof(BlockNumber) * LDB_HNSW_INSERT_MAX_EXTRA_DIRTIED_BUFS);
+    ed->EXTRA_DIRTIED_BUF = palloc0(sizeof(Buffer) * LDB_HNSW_INSERT_MAX_EXTRA_DIRTIED_BUFS);
+    ed->EXTRA_DIRTIED_PAGE = palloc0(sizeof(Page) * LDB_HNSW_INSERT_MAX_EXTRA_DIRTIED_BUFS);
+    ed->EXTRA_DIRTIED_SIZE = 0;
+    return ed;
+}
+
+void extra_dirtied_add(ExtraDirtiedBufs *ed, BlockNumber blockno, Buffer buf, Page page)
```

Comment:
maybe not good suggestion, but I'd rather pass the context and then get the buffers from the context here